torch/csrc/jit/passes/memory_planning.cpp

```diff
@@ -0,0 +1,399 @@
+  // TODO: when does this fail? answer: in place mutation
+  auto numel = ttp->numel();
+  if (!numel.has_value()) {
+    TORCH_WARN(
+        node_header, "; ", value.debugName(), " doesn't have numel: ", *ttp);
+    return c10::nullopt;
+  }
+
+  return numel.value() * element_size;
+}
+
+Node* insertSlabNode(
+    const std::shared_ptr<Graph>& graph,
+    size_t total_size,
+    c10::optional<at::Device> device_type = c10::nullopt) {
+  auto slab = graph->create(prim::AllocateSlab, 1);
+  slab->output()->setType(StorageType::get());
+  slab->i_(attr::total_size, (int64_t)total_size);
+
+  if (device_type || (device_type = jit::tensorexpr::pickDeviceType(graph))) {
+    slab->i_(attr::device_type, static_cast<int8_t>(device_type->type()));
+  } else {
+    slab->i_(attr::device_type, static_cast<int8_t>(at::kCPU));
+  }
+  slab->insertBefore(graph->nodes().front());
+  return slab;
+}
+
+void insertAllocNodes(
+    const std::shared_ptr<Graph>& graph,
+    size_t total_size,
+    std::vector<std::pair<const Value*, MemRegion>>& managed_value_allocations,
+    c10::optional<at::Device> device_type = c10::nullopt) {
+  auto slab = insertSlabNode(graph, total_size, device_type);
+  auto release_slab = graph->create(prim::ReleaseSlab, 0);
+  release_slab->addInput(slab->output());
+
+  for (auto& item : managed_value_allocations) {
+    auto val = item.first;
+    auto region = item.second;
+
+    // const_cast fishy?
+    Node* node = const_cast<Node*>(val->node());
+
+    // the way that this node magically *becomes* the out varaint is simply
+    // by add an extra input. this is because op resolution happens
+    // at runtime via the op registry (by matching on the schema).
+    auto* alloc = graph->create(prim::AllocateTensor, 1);
+    node->addInput(alloc->output());
+    alloc->insertBefore(node);
+    alloc->addInput(slab->output());
+
+    auto ttp = val->type()->expect<c10::TensorType>();
+    alloc->ty_(attr::profiled_type, ttp);
+    alloc->i_(attr::size, (int64_t)region.size);
+    alloc->i_(attr::offset, (int64_t)region.offset);
+    release_slab->addInput(alloc->output());
+  }
+  release_slab->insertBefore(graph->return_node());
+}
+
+bool hasOutVariant(Node* node) {
+  if (!node->maybeSchema()) {
+    return false;
+  }
+  const auto& node_schema_args = node->schema().arguments();
+  for (const auto& variant : getAllOperatorsFor(node->kind())) {
+    auto variant_schema = variant->schema();
+    // a simple check for an out param is insufficient because ops with multiple
+    // out variants e.g. aten::cat.out and aten::cat.names_out
+    auto maybe_out_arg_idx = variant_schema.argumentIndexWithName("out");
+    if (maybe_out_arg_idx) {
+      // out arg should be the *next* arg after current args
+      if ((size_t)maybe_out_arg_idx.value() != node_schema_args.size()) {
+        return false;
+      }
+      // functions are contravariant in arguments i.e. args should be "broader"
```

Comment:
Nice!! Was this taken from anywhere ? BC incompat test maybe ? Worth refactoring?

Reply:
lol i lifted it straight out of [`FunctionSchema::isSubtypeOf`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/function_schema_inl.h#L317) which differs only in that it checks *all* args (whereas i'm checking only up until the `out` input)